mpcs/src/basefold/commit_phase.rs

```diff
@@ -86,97 +90,108 @@ where
-                    p3::matrix::dense::DenseMatrix::new(evaluations, 1),
-                );
-                assert_eq!(basecode.values, new_running_oracle.values);
-            }
-        }
-        end_timer!(sumcheck_timer);
+
+    let sumcheck_phase1 = entered_span!("sumcheck_phase1");
+    for i in 0..num_rounds.min(num_vars - log2_strict_usize(num_threads)) {
+        challenge = basefold_one_round::<E, Spec>(
+            pp,
+            &mut prover_states,
+            challenge,
+            &mut sumcheck_messages,
+            &initial_oracle,
+            transcript,
+            &mut trees,
+            &mut roots,
+            &mmcs_ext,
+            i == num_rounds - 1,
+        );
+    }
+    exit_span!(sumcheck_phase1);
+
+    if let Some(p) = challenge {
+        prover_states
+            .iter_mut()
+            .for_each(|prover_state| prover_state.fix_var(p.elements));
+    }
+
+    // deal with log(#thread) basefold rounds
+    let merge_sumcheck_polys_span = entered_span!("merge_sumcheck_polys");
+    let poly = merge_sumcheck_polys(&prover_states);
+    let mut prover_states = vec![IOPProverState::prover_init_with_extrapolation_aux(
+        poly,
+        vec![(vec![], vec![])],
+    )];
+    exit_span!(merge_sumcheck_polys_span);
+
+    let mut challenge = None;
+
+    let sumcheck_phase2 = entered_span!("sumcheck_phase2");
+    for i in 0..num_rounds.saturating_sub(num_vars - log2_strict_usize(num_threads)) {
+        challenge = basefold_one_round::<E, Spec>(
+            pp,
+            &mut prover_states,
+            challenge,
+            &mut sumcheck_messages,
+            &initial_oracle,
+            transcript,
+            &mut trees,
+            &mut roots,
+            &mmcs_ext,
+            i == num_rounds.saturating_sub(num_vars - log2_strict_usize(num_threads)) - 1,
+        );
+    }
+    exit_span!(sumcheck_phase2);
+
+    if let Some(p) = challenge {
+        prover_states[0].fix_var(p.elements);
     }
-    end_timer!(timer);
+
+    let mut final_message = prover_states[0].get_mle_final_evaluations();
+    // skip first half which is eq evaluations
+    let final_message = final_message.split_off(final_message.len() / 2);
+
+    if cfg!(feature = "sanity-check") {
+        // If the prover is honest, in the last round, the running oracle
+        // on the prover side should be exactly the encoding of the folded polynomial.
+        let basecode = <Spec::EncodingScheme as EncodingScheme<E>>::encode_slow_ext(
+            p3::matrix::dense::DenseMatrix::new(final_message.clone(), 1),
+        );
+        assert_eq!(
+            basecode.values,
+            mmcs_ext.get_matrices(trees.last().unwrap())[0].values
+        );
+        // remove last tree/commmitment which is only for debug purpose
+        let _ = (trees.pop(), roots.pop());
+    }
+    transcript.append_field_element_exts(&final_message);
```

Comment:
why not write eq evaluations to transcript?

Reply:
since verifier can derive eq deterministic by himself